rp2040-hal/src/multicore.rs

```diff
@@ -290,3 +331,68 @@ impl<'p> Core<'p> {
         }
     }
 }
+
+#[cfg(all(target_arch = "arm", feature = "thread_local"))]
+mod thread_local {
+    use core::arch::global_asm;
+    use core::ptr::addr_of;
+
+    extern "C" {
+        static TLS_CORE_0: u8;
+        static TLS_CORE_1: u8;
+    }
+    // Not really a const pointer, but we reform it into mut in the asm
+    static mut TLS_STATE: [*const u8; 2] = [
+        // Point to linker-allocated space in .bss
+        unsafe { addr_of!(TLS_CORE_0) },
+        unsafe { addr_of!(TLS_CORE_1) },
+    ];
+
+    // Define `__aeabi_read_tp` called by the compiler to get access to
+    // thread-local storage.
+    global_asm! {
+        ".pushsection .text.__aeabi_read_tp",
+        ".align 4",
+        ".p2align 4,,15",
+        ".global __aeabi_read_tp",
+        ".type __aeabi_read_tp,%function",
+
+        "__aeabi_read_tp:",
+        "    push {{r1, lr}}",
+        "    ldr r1, =0xd0000000",      // Load SIO CPUID addr
+        "    ldr r1, [r1]",             // Get current CPUID
+        "    lsls r1, r1, #2",          // Scale by 4
+        "    ldr r0, ={tls_state}",     // Load TLS_STATE base addr
+        "    ldr r0, [r0, r1]",         // Load CPU per-thread
+        "    pop {{r1, pc}}",
+
+        ".popsection",
+        tls_state = sym TLS_STATE,
+    }
+
+    // Intercept __pre_init to hook into the startup code to copy the tdata into
+    // TLS_CORE_[01].
+    global_asm! {
+        ".pushsection .text.__pre_init",
+        ".align 4",
+        ".p2align 4,,15",
+        ".global __pre_init",
+        ".type __pre_init,%function",
+
+        "__pre_init:",
+        "    push {{lr}}",
+        "    ldr r0, ={tls_core_0}",
+        "    ldr r1, =__tdata_start",
+        "    ldr r2, =__tdata_len",
+        "    bl __aeabi_memcpy",
+        "    ldr r0, ={tls_core_1}",
+        "    ldr r1, =__tdata_start",
+        "    ldr r2, =__tdata_len",
+        "    bl __aeabi_memcpy",
+        "    pop {{pc}}",
+
+        ".popsection",
+        tls_core_0 = sym TLS_CORE_0,
+        tls_core_1 = sym TLS_CORE_1,
+    }
```

Comment:
Why not using a regular rust funtion placed in that section with the [`link_section` attribute](https://doc.rust-lang.org/reference/abi.html#the-link_section-attribute) ?

EG:
```rust
#[no_mangle]
#[link_section = ".text.__pre_init"]
unsafe fn tls_pre_init_hook() {
    for dts in [ addr_of!(TLS_CORE_0), addr_of!(TLS_CORE_1) ] {
        core::ptr::copy(addr_of!(__tdata_start), dst, addr_of!(__tdata_len));
    }
}
```

Reply:
I don't think encouraging people to put rust code in pre-init is a good idea (in general) - [there are warnings on the cortex-m pre_init macro for a reason](https://docs.rs/cortex-m-rt/latest/cortex_m_rt/attr.pre_init.html#safety).
rust-embedded moved away from having r0 written in rust for this same reason - it is easier to conform with conditions required pre-init in asm than it is in rust.

Reply:
Yeah I wasn't sure about whether Rust would be valid that early. Though the code using `addr_of` avoids forming any reference which could cause problems. And maybe sprinkling some `MaybeUninit` would make things a bit more robust.